File: solutions/538.convert-bst-to-greater-tree/convert-bst-to-greater-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var convertBST = function(root) {
    var sum = 0;
    function helper(root){
        if(root === null) return null;
        helper(root.right);
        root.val += sum;
        sum = root.val;
        helper(root.left);
        return root;
    }
    return helper(root);
};

File: solutions/538.convert-bst-to-greater-tree/convert-bst-to-greater-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode convertBST(TreeNode root) {
        int sum = 0;
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode cur = root;
        while(cur != null || !stack.isEmpty()){
            if(cur != null){
                stack.push(cur);
                cur = cur.right;
            }else{
                TreeNode node = stack.pop();
                node.val += sum;
                sum = node.val;
                cur = node.left;
            }
        }
        return root;
    }
}